File: ED del Pedro/ED/Java/Examenes ED Java/Febrero 2016/Febrero 2016/src/dataStructures/graph/EulerianCycle.java
/**
 * Student's name:
 * Student's group:
 *
 * Data Structures. Grado en Informática. UMA.
 */

package dataStructures.graph;

import dataStructures.list.*;
import dataStructures.set.HashSet;
import dataStructures.set.Set;

public class EulerianCycle<V> {
    private List<V> eCycle;

    @SuppressWarnings("unchecked")
    public EulerianCycle(Graph<V> g) {
        Graph<V> graph = (Graph<V>) g.clone();
        eCycle = eulerianCycle(graph);
    }

    public boolean isEulerian() {
        return eCycle != null;
    }

    public List<V> eulerianCycle() {
        return eCycle;
    }

    // J.1
    private static <V> boolean isEulerian(Graph<V> g) {

        boolean isEul = true;

        for (V v : g.vertices()){
            if(g.degree(v) % 2 != 0){
                isEul = false;
                break;
            }
        }

        return isEul;
    }

    // J.2
    private static <V> void remove(Graph<V> g, V v, V u) {

        g.deleteEdge(v, u);
        if (g.degree(v) == 0){
            g.deleteVertex(v);
        }
        if (g.degree(u) == 0){
            g.deleteVertex(u);
        }

    }

    // J.3
    private static <V> List<V> extractCycle(Graph<V> g, V v0) {
        List<V> cycle = new LinkedList<>();
        cycle.append(v0);
        Set<V> suc = g.successors(v0);
        V v = v0;

        while (! suc.isEmpty()){
            V u = suc.iterator().next();
            cycle.append(u);
            remove(g, v, u);
            v = u;
            suc = g.successors(v);
        }

        return cycle;
    }

    // J.4
    private static <V> void connectCycles(List<V> xs, List<V> ys) {

        if (xs.isEmpty()){
            xs = ys;
        }else {
            int i = 0;
            boolean found = false;
            while (i < xs.size() && !found){
                found = xs.get(i) == ys.get(0);
                i ++;
            }

            for (int j = 1; j < ys.size(); j ++){
                V v = ys.get(j);
                xs.insert(i, v);
                i ++;
            }

        }

    }

    // J.5
    private static <V> V vertexInCommon(Graph<V> g, List<V> cycle) {

        V vic = cycle.get(0);

        if (!g.vertices().isElem(vic)){
            for (V v : cycle){
                if (g.vertices().isElem(v)){
                    vic = v;
                    break;
                }
            }
        }

        return vic;
    }

    // J.6
    private static <V> List<V> eulerianCycle(Graph<V> g) {

        List<V> cycle = null;

        if (isEulerian(g)){
            V v = g.vertices().iterator().next();
            cycle = extractCycle(g, v);
            // extractCycle(g, v);

            while (!g.isEmpty()){
                V vic = vertexInCommon(g, cycle);
                List<V> cycle2 = extractCycle(g, vic);
                // extractCycle(g, vic);
                connectCycles(cycle, cycle2);

            }
        }

        return cycle;
    }
}

File: ED del Pedro/Ejercicios Haskell/Examenes Estructura de Datos/haskell-feb-2016/haskell-feb-2016/haskell/DataStructures/Graph/EulerianCycle.hs
-------------------------------------------------------------------------------
-- Student's name:
-- Student's group:
--
-- Data Structures. Grado en Informática. UMA.
-------------------------------------------------------------------------------

module DataStructures.Graph.EulerianCycle(isEulerian, eulerianCycle) where

import DataStructures.Graph.Graph
import Data.List

--H.1)
isEulerian :: Eq a => Graph a -> Bool
isEulerian g = aux (vertices g)
    where 
        aux [] = True
        aux (v:vs) = mod (degree g v) 2 == 0 && aux vs

-- H.2)
remove :: (Eq a) => Graph a -> (a,a) -> Graph a
remove g (v,u) = aux (deleteEdge g (v,u)) (vertices g)
    where
        aux g' [] = g'
        aux g' (v:vs) = if degree g' v == 0 then aux (deleteVertex g' v) vs else aux g' vs

-- H.3)

modify :: Eq a => Graph a -> Path a -> Graph a
modify g (x:y:xs) = if edges g /= [] then modify (remove g (x,y)) (y:xs) else mkGraphEdges (vertices g) []
modify empty _ = empty

searchCycle :: Eq a => Graph a -> a -> a -> [a] -> [a] -> Path a
searchCycle g' v v0 (suc:resSucs) sol = if suc /= v0 then sol ++ [suc] ++ searchCycle g' suc v0 (successors g' suc) sol else sol ++ [v0]

extractCycle :: (Eq a) => Graph a -> a -> (Graph a, Path a)
extractCycle g v0 = aux g v0 (successors g v0) [v0]
    where
        aux g v [] c = (g, c)
        aux g v (u : us) c = aux newG u (successors newG u) (c ++ [u])
            where 
                newG = remove g (v, u)

-- H.4)
connectCycles :: (Eq a) => Path a -> Path a -> Path a
connectCycles [] ys' = ys'
connectCycles xs (y:ys) = if head xs == y then (y:ys) ++ tail xs else head xs:connectCycles (tail xs) (y:ys)

-- H.5)
vertexInCommon :: Eq a => Graph a -> Path a -> a
vertexInCommon g cycle = aux (vertices g) cycle
    where 
        aux (v:vs) cycle = if v `elem` cycle then v else aux vs cycle

-- H.6) 
eulerianCycle :: Eq a => Graph a -> Path a
eulerianCycle g = if isEulerian g then aux (extractCycle g (head (vertices g))) else error "The graph is not eulerian"
    where
        aux (g, c)
            | isEmpty g = c
            | otherwise = aux (newG, (connectCycles c newC))
            where
                newG = fst (extractCycle g (vertexInCommon g c))
                newC = snd (extractCycle g (vertexInCommon g c))
